File: ruby/feed-forward-iris-softmax-cross-entropy.rb
#!/usr/bin/ruby -w

require "numo/narray"
require 'csv'

def sigmoid(x, grad)
    if grad
        return sigmoid(x, false) * (1 - sigmoid(x, false))
    else
        return 1 / (1 + Numo::NMath.exp((x * -1)))
    end
end

def softmax(x, grad)
    if grad
        p = softmax(x, false)
        return p * (1-p)
    else
        e_x = Numo::NMath.exp(x - x.max(axis:1,keepdims:true))
        return e_x / e_x.sum(axis:1,keepdims:true)
    end
end

def cross_entropy(x, y, grad)
    if grad
        x = x.clip(1e-15, 1-1e-15)
        return -(y / x) + (1 - y) / (1 - x)
    else
        x = x.clip(1e-15, 1-1e-15)
        return -y * Numo::NMath.log(x) - (1 - y) * Numo::NMath.log(1 - x)
    end
end

def accuracy(x, y)
    correct = 0
    $i = 0
    while $i < x.shape[0]
        if x[$i] == y[$i]
            correct += 1
        end
        $i +=1
    end
    return (correct * 1.0) / x.shape[0]
end

# read our csv
csv = CSV.read('iris.csv', :headers=>true)
sepal_length = csv['SepalLengthCm'].map(&:to_f)
sepal_width = csv['SepalWidthCm'].map(&:to_f)
petal_length = csv['PetalLengthCm'].map(&:to_f)
petal_width = csv['PetalWidthCm'].map(&:to_f)
flowers = csv['Species']
unique_flowers = flowers.uniq

# change from ruby array to numo array
iris = Numo::NArray.[](sepal_length, sepal_width, petal_length, petal_width).transpose

# our Y and onehot-encoder
flowers_int = Array.new(flowers.length)
onehot = Numo::DFloat.zeros(flowers.length, unique_flowers.length)
$i = 0
while $i < flowers.length do
    flowers_int[$i] = unique_flowers.find_index(flowers[$i])
    onehot[$i, unique_flowers.find_index(flowers[$i])] = 1
    $i +=1
end

# our global variables
size_layer = 64
learning_rate = 1e-4
epoch = 50

w1 = Numo::DFloat.new(iris.shape[1], size_layer).rand
b1 = Numo::DFloat.new(size_layer).rand
w2 = Numo::DFloat.new(size_layer, size_layer).rand
b2 = Numo::DFloat.new(size_layer).rand
w3 = Numo::DFloat.new(size_layer, unique_flowers.length).rand
b3 = Numo::DFloat.new(unique_flowers.length).rand

iteration = 0
while iteration < epoch
    a1 = iris.dot(w1) + b1
    z1 = sigmoid(a1,false)
    a2 = z1.dot(w2) + b2
    z2 = sigmoid(a2,false)
    a3 = z2.dot(w3) + b3
    y_hat = softmax(a3,false)
    acc = accuracy(y_hat.max_index(1), onehot.max_index(1))
    cost = cross_entropy(y_hat,onehot, false).mean()
    dy_hat = cross_entropy(y_hat,onehot, true)
    da3 = softmax(a3, true) * dy_hat
    dw3 = z2.transpose.dot(da3)
    db3 = da3.sum(0)
    dz2 = da3.dot(w3.transpose)
    da2 = sigmoid(a2, true) * dz2
    dw2 = z1.transpose.dot(da2)
    db2 = da2.sum(0)
    dz1 = da2.dot(w2.transpose)
    da1 = sigmoid(a1, true) * dz1
    dw1 = iris.transpose.dot(da1)
    db1 = da1.sum(0)
    w3 += -learning_rate * dw3
    b3 += -learning_rate * db3
    w2 += -learning_rate * dw2
    b2 += -learning_rate * db2
    w1 += -learning_rate * dw1
    b1 += -learning_rate * db1
    p "epoch #{iteration+1}, loss #{cost}, accuracy #{acc}"
    iteration += 1
end

File: julia/feed-forward-iris-softmax-cross-entropy.jl
using DataFrames, CSV

function softmax(X, grad)
    if grad
        p = softmax(X,false)
        return p .* (1-p)
    else
        maxs = findmax(X,2)[1]
        maxs = repeat(maxs, outer=[1,size(X)[2]])
        X = exp(X - maxs)
        sums = sum(X,2)
        sums = repeat(sums, outer=[1,size(X)[2]])
        return X ./ sums
    end
end

function sigmoid(X, grad)
    if grad
        return sigmoid(X,false) .* (1-sigmoid(X,false))
    else
        return 1 ./ (1 - exp(-X))
    end
end

function cross_entropy(X, Y, grad)
    if grad
        X = clamp(X, 1e-15, 1 - 1e-15)
        return -(Y ./ X) + (1 - Y) ./ (1 - X)
    else
        X = clamp(X, 1e-15, 1 - 1e-15)
        return -Y .* log(X) - (1 - Y) .* log(1 - X)
    end
end

iris = readtable("iris.csv",separator=',')
X = convert(Array, iris[:,2:size(iris)[2]-1])
Y = convert(Array, iris[:,size(iris)[2]])
unique_flowers = unique(Y)

flowers = Int64[]
for i in 1:size(X)[1]
    push!(flowers,find(unique_flowers .== Y[i])[1])
end

onehot = zeros((size(X)[1],size(unique_flowers)[1]))
for i in 1:size(X)[1]
    onehot[i, flowers[i]] = 1
end

EPOCH = 100
LEARNING_RATE = 0.0005
W1 = randn((size(X)[2],64))
W2 = randn((64,64))
W3 = randn((64,size(unique_flowers)[1]))

for i in 1:EPOCH
    a1 = X * W1
    z1 = sigmoid(a1,false)
    a2 = z1 * W2
    z2 = sigmoid(a2,false)
    a3 = z2 * W3
    y_hat = softmax(a3,false)
    accuracy = mean(findmax(y_hat,2)[2] .== findmax(onehot,2)[2])
    cost = mean(cross_entropy(y_hat,onehot,false))
    dy_hat = cross_entropy(y_hat,onehot,true)
    da3 = softmax(a3,true) .* dy_hat
    dW3 = transpose(z2) * da3
    dz2 = da3 * transpose(W3)
    da2 = sigmoid(a2, true) .* dz2
    dW2 = transpose(z1) * da2
    dz1 = da2 * transpose(W2)
    da1 = sigmoid(a1, true) .* dz1
    dW1 = transpose(X) * da1
    W3 += -LEARNING_RATE * dW3
    W2 += -LEARNING_RATE * dW2
    W1 += -LEARNING_RATE * dW1
    @printf("epoch %d, cost %f, accuracy %f\n",i, cost, accuracy)
end
